notifications/v1/client.go

```diff
@@ -6,90 +6,153 @@ import (
-// notifications. If the handler of a notification is nil, the notification will
-// be acknowledged automatically without any processing.
-//
-// If a handler returns an error, the notification will be unacknowledged and
-// retried later.
-type ReceiveHandlers struct {
-	// OnUserDeleted is called when a "UserDeleted" notification is received.
-	OnUserDeleted func(data *UserDeletedData) error
-}
+func (s *subscriber) Start() {
+	if !s.state.CompareAndSwap(stateIdle, stateStarted) {
+		s.logger.Error("failed to start subscriber", log.Error(errors.New(`not in "idle" state`)))
+		return
+	}
 
-// Receive starts receiving SAMS notifications and calls the corresponding
-// handler for each notification. It blocks until the context is done (e.g.
-// deadline exceed or canceled).
-//
-// ⚠️ WARNING: Each subscription can only have one active receiver at a time,
-// i.e. there should only be one call to Receive for a given subscription.
-func (c *Client) Receive(handler *ReceiveHandlers) error {
 	ctx, cancel := context.WithCancel(context.Background())
-	c.cancelContext = cancel
-	return c.subscription.Receive(ctx, func(ctx context.Context, msg *pubsub.Message) {
+	s.cancelContext = cancel
+	err := s.subscription.Receive(ctx, func(ctx context.Context, msg *pubsub.Message) {
 		var msgData struct {
 			Name     string          `json:"name"`
 			Metadata json.RawMessage `json:"metadata"`
 		}
 		err := json.Unmarshal(msg.Data, &msgData)
 		if err != nil {
-			c.logger.Error("failed to unmarshal notification message", log.Error(err))
+			s.logger.Error("failed to unmarshal notification message", log.Error(err))
 			msg.Nack()
 			return
 		}
 
-		err = c.handleReceive(handler, msgData.Name, msgData.Metadata)
-		if err != nil {
-			c.logger.Error("failed to process notification message", log.Error(err))
+		err = s.handleReceive(msgData.Name, msgData.Metadata)
+		if err == nil {
+			msg.Ack()
+		} else {
+			s.logger.Error("failed to process notification message", log.Error(err))
 			msg.Nack()
-			return
 		}
-		msg.Ack()
 	})
+	if err != nil {
+		s.logger.Error("failed to receive notifications", log.Error(err))
+		return
+	}
+}
+
+func (r *subscriber) Stop() {
+	if !r.state.CompareAndSwap(stateStarted, stateStopped) {
+		r.logger.Error("failed to stop subscriber", log.Error(errors.New(`not in "started" state`)))
+		return
+	}
+
+	r.cancelContext()
+	r.logger.Info("subscriber stopped")
 }
 
-func (c *Client) handleReceive(handler *ReceiveHandlers, name string, metadata json.RawMessage) error {
+// SubscriberHandlers is a collection of subscription handlers for each type of
+// SAMS notifications. If the handler of a notification is nil, the notification
+// will be acknowledged automatically without any processing.
+//
+// If a handler returns an error, the notification will be unacknowledged and
+// retried later.
+type SubscriberHandlers struct {
+	// OnUserDeleted is called when a "UserDeleted" notification is received.
```

Comment:
Where is the "most canonical" documentation for this that you expect people who integrate with this code?

If we expect that people who are wiring up a `SubscriptionHandler` for a SAMS-dependent service, I would suggest we try to expand on the doc comments here to help explain when/why that "UserDeleted" notification would be received.

... not that it is confusing in this particular situation, but I'd nudge you to do that in this first commit so that we will follow that pattern as we add new events. Because when the event is more ambiguous, e.g. `OnUserProfileUpdated` or `OnAccountVerified`, putting a 1-2 sentences explanation would make it easier for people to integrate with this correctly. (e.g. understanding what to expect and not-expect when they see this event.)

Reply:
Enriched the docstring in https://github.com/sourcegraph/sourcegraph-accounts-sdk-go/pull/24/commits/e1a5374399524832b1f3444c034e3b40c1687d00!